File: src/main/data/thumbnailDataStore.ts
import path from 'path'
import fs from 'fs'
import fsPromises from 'fs/promises'
import crypto from 'crypto'

export class ThumbnailDataStore {
  cacheFolder: string

  constructor(userDataPath: string) {
    this.cacheFolder = path.join(userDataPath, 'thumbnails')

    if (!fs.existsSync(this.cacheFolder)) {
      fs.mkdirSync(this.cacheFolder)
    }
  }

  private async ensurePlaylistFolder(playlistId: string): Promise<string> {
    const folder = path.join(this.cacheFolder, playlistId)

    try {
      await fsPromises.mkdir(folder, { recursive: true })
    } catch (err: any) {
      // If the error is "File exists", we don't care. It's a success for us.
      if (err.code !== 'EEXIST') {
        throw err // Rethrow actual errors (permissions, disk full, etc.)
      }
    }
    return folder
  }

  private static getThumbnailPathWithPlaylistPath(playlistPath: string, playlistItemId: string, fileHash: string): string {
    return path.join(playlistPath, `${playlistItemId}-${fileHash}.jpeg`)
  }

  private async getThumbnailPath(playlistId: string, playlistItemId: string, fileHash: string): Promise<string> {
    return ThumbnailDataStore.getThumbnailPathWithPlaylistPath(await this.ensurePlaylistFolder(playlistId), playlistItemId, fileHash)
  }

  public calculateFileHash(filePath: string, fileSize: number, lastModified: number): string {
    const input = `${filePath}-${fileSize}-${lastModified}`
    return crypto.createHash('sha256').update(input).digest('hex')
  }

  public async getThumbnail(playlistId: string, playlistItemId: string, fileHash: string): Promise<string | null> {
    const thumbPath = await this.getThumbnailPath(playlistId, playlistItemId, fileHash)
    const stats = await fsPromises.stat(thumbPath).catch(() => null)

    if (!stats || !stats.isFile()) {
      return null
    }

    try {
      const imageBuffer = await fsPromises.readFile(thumbPath)
      return imageBuffer.toString('base64')
    } catch (e) {
      console.error("Failed to read thumbnail:", e)
      return null
    }
  }

  public async setThumbnail(playlistId: string, playlistItemId: string, fileHash: string, imageBuffer: Buffer) {
    const playlistFolder = await this.ensurePlaylistFolder(playlistId)

    const files = await fsPromises.readdir(playlistFolder)
    const prefix = `${playlistItemId}-`

    const deletePromises = files
      .filter(file => file.startsWith(prefix) && file.endsWith('.jpeg'))
      .map(file => fsPromises.unlink(path.join(playlistFolder, file)).catch(() => {
        // Silently catch errors (e.g., file already deleted)
      }))

    await Promise.all(deletePromises)

    const newThumbPath = ThumbnailDataStore.getThumbnailPathWithPlaylistPath(playlistFolder, playlistItemId, fileHash)
    await fsPromises.writeFile(newThumbPath, imageBuffer)
  }
}

File: internal/data/thumbnail_data_store.go
package data

import (
	"crypto/sha256"
	"encoding/base64"
	"fmt"
	"os"
	"path/filepath"
)

type ThumbnailDataStore struct {
	cacheFolder string
}

func NewThumbnailDataStore(userDataPath string) *ThumbnailDataStore {
	folder := filepath.Join(userDataPath, "thumbnails")
	if err := os.MkdirAll(folder, 0755); err != nil {
		panic(fmt.Sprintf("failed to create thumbnail cache dir: %v", err))
	}
	return &ThumbnailDataStore{cacheFolder: folder}
}

func (s *ThumbnailDataStore) ensurePlaylistFolder(playlistID string) (string, error) {
	folder := filepath.Join(s.cacheFolder, playlistID)
	if err := os.MkdirAll(folder, 0755); err != nil {
		return "", fmt.Errorf("failed to create playlist thumbnail folder: %w", err)
	}
	return folder, nil
}

func thumbnailPath(playlistPath, itemID, fileHash string) string {
	return filepath.Join(playlistPath, fmt.Sprintf("%s-%s.jpeg", itemID, fileHash))
}

func (s *ThumbnailDataStore) CalculateFileHash(filePath string, fileSize int64, lastModified int64) string {
	input := fmt.Sprintf("%s-%d-%d", filePath, fileSize, lastModified)
	hash := sha256.Sum256([]byte(input))
	return fmt.Sprintf("%x", hash)
}

func (s *ThumbnailDataStore) GetThumbnail(playlistID, itemID, fileHash string) (string, error) {
	playlistPath, err := s.ensurePlaylistFolder(playlistID)
	if err != nil {
		return "", err
	}

	thumbPath := thumbnailPath(playlistPath, itemID, fileHash)
	stats, err := os.Stat(thumbPath)
	if err != nil || stats.IsDir() {
		return "", nil
	}

	data, err := os.ReadFile(thumbPath)
	if err != nil {
		return "", nil
	}

	return fmt.Sprintf("data:image/jpeg;base64,%s", base64.StdEncoding.EncodeToString(data)), nil
}

func (s *ThumbnailDataStore) SetThumbnail(playlistID, itemID, fileHash string, data []byte) error {
	playlistPath, err := s.ensurePlaylistFolder(playlistID)
	if err != nil {
		return err
	}

	entries, err := os.ReadDir(playlistPath)
	if err == nil {
		prefix := fmt.Sprintf("%s-", itemID)
		for _, entry := range entries {
			if !entry.IsDir() && len(entry.Name()) > len(prefix) && entry.Name()[:len(prefix)] == prefix && filepath.Ext(entry.Name()) == ".jpeg" {
				os.Remove(filepath.Join(playlistPath, entry.Name()))
			}
		}
	}

	newPath := thumbnailPath(playlistPath, itemID, fileHash)
	return os.WriteFile(newPath, data, 0644)
}
